onnx/defs/nn/defs.cc

```diff
@@ -3156,15 +3156,36 @@ ONNX_OPERATOR_SET_SCHEMA(
           auto* num_heads_attr = ctx.getAttribute("num_heads");
           int64_t num_heads = (num_heads_attr != nullptr) ? num_heads_attr->i() : 0;
 
+          // Ensure that num_heads does not control reshaping of input tensor
+          // when input tensor is 4D
+          int64_t is_input_4d = 1;
+          auto* x_tp = ctx.getInputType(0);
+          if ((x_tp == nullptr) || (!x_tp->has_tensor_type()))
+            return false;
+          if (!(x_tp->tensor_type().has_shape())) {
+            return false;
+          }
+          if (x_tp->tensor_type().shape().dim_size() == 4) {
+            num_heads = 0;
```

Comment:
The above assignment to num_heads may be unnecessary if the reshape (for 3D) is moved into the if-branch as discussed below?